code to ingest real-time weather data as CSVs in S3

Retr0Metal98 opened the pull request:
Wrote a script to run through the data.gov.sg weather APIs and ingest the real-time data (will update later to include functions to collect weather forecasts and pollution data)

Will need to refactor data format as needed by model team, and for ingestion into AWS S3 and for use by the spark system to accept.

Retr0Metal98 committed:
code to collect real-time weather data into pandas dataframes, will need to refactor data format as needed by model team, and for ingestion into AWS S3 and for use by the spark system to accept.

don-tay reviewed (commented):
For the file `lta_camera.py`, it is essentially what `data_ingest.py` is already doing. The main difference is that lta_camera.py stores the file locally while `data_ingest.py` uploads this same file to S3.

Would you like the local file storage functionality to be implemented as well?

don-tay commented on weather_data.py:
```diff
@@ -0,0 +1,76 @@
+real_time_weather_dfs = real_time_weather()
+for (field, df) in real_time_weather_dfs.items():
+    print("Field:", field)
+    print(df)
```
Should we extend this script to store the results into a relational db?

Retr0Metal98 commented:
I guess we can keep the local file storage functionality to use for quick testing purposes? Perhaps we can merge the two files and let the user toggle between AWS storage/local storage somehow?

Btw, I think don't merge the branches yet, I can add more functions for the other data sources (weather forecast, pollution data, etc.)

don-tay commented:
> I guess we can keep the local file storage functionality to use for quick testing purposes? Perhaps we can merge the two files and let the user toggle between AWS storage/local storage somehow?
> 
> Btw, I think don't merge the branches yet, I can add more functions for the other data sources (weather forecast, pollution data, etc.)

Yes this is possible, we can use an env flag to toggle between `local` and `development`. That said, I suggest we see what works best for the development flow in the model team

Retr0Metal98 committed:
Merge branch 'master' of https://github.com/don-tay/traffic-predict into weather-data-ingestion

Retr0Metal98 committed:
Update weather_data to upload CSVs to S3

Retr0Metal98 committed:
2-hr weather forecast ingestion, bucket save and merge files

function to ingest 2-hr weather forecast data as CSVs

code to output csv files to S3 bucket/local filesystem

code to merge timestamped csv files and reupload to S3.

job scheduling code added

Retr0Metal98 committed:
Completed weather data ingestion functions

Code for ingesting real-time data, forecast data (2-hr, 24-hr, 4-day) completed.

CSV file writing and merging operations moved to helper.py

Ingestion job scheduling moved to task_schedule.py

don-tay commented on helper.py:
```diff
@@ -32,9 +45,91 @@ def get_req_handler(url: str) -> requests.Response:
+            except Exception:
+                print("Failed to upload data in " + savepath, file=sys.stderr)
+            else:
+                print("Uploaded dataframe as csv file " + savepath + " to bucket!")
```
Shortening this portion with the same effect
```suggestion
                s3_bucket.put_object(Body=csv_buffer.getvalue(), Key=savepath)
                print("Uploaded dataframe as csv file " + savepath + " to bucket!")
            except Exception:
                print("Failed to upload data in " + savepath, file=sys.stderr)
```

don-tay commented on helper.py:
```diff
@@ -32,9 +45,91 @@ def get_req_handler(url: str) -> requests.Response:
+            else:
+                print("Uploaded dataframe as csv file " + savepath + " to bucket!")
+        else:
+            print("When using output='AWS', specify S3.Bucket object in s3_bucket.")
```
Cutting down on nested conditions with guard clause
```suggestion
        if not s3_bucket:
            print("When using output='AWS', specify S3.Bucket object in s3_bucket.")
            return
            
              try:
                  # upload dataframe to S3 bucket
                  csv_buffer = StringIO()
                  dataframe.to_csv(csv_buffer, encoding="utf-8")
                 s3_bucket.put_object(Body=csv_buffer.getvalue(), Key=savepath)
             except Exception:
                 print("Failed to upload data in " + savepath, file=sys.stderr)
```

don-tay commented on lta_camera.py:
```diff
@@ -0,0 +1,49 @@
+import os
```
We can delete this file since `data_ingest.py` can be extended to include local file upload

don-tay commented on lta_camera.py:
```diff
@@ -0,0 +1,49 @@
+
+        with open(img_file_path, 'xb') as handler:
+            handler.write(img_data)
+        handler.close()
```
When `with open(...)...` is used, the file `close()` will be handled automatically. Hence there is no need to call the `close()` method
```suggestion
        with open(img_file_path, 'xb') as handler:
            handler.write(img_data)
```

don-tay reviewed (approved):
LGTM, some code refactor, and i think we can do away with `lta_camera.py` and we are good to go

Retr0Metal98 committed:
Update helper.py

Co-authored-by: Don Tay <dontay0209@gmail.com>

Retr0Metal98 committed:
Update lta_camera.py

Co-authored-by: Don Tay <dontay0209@gmail.com>

Retr0Metal98 committed:
Update helper.py

included suggestion in PR #1 : Cutting down on nested conditions with guard clause

Retr0Metal98 committed:
Merge branch 'weather-data-ingestion' of https://github.com/don-tay/traffic-predict into weather-data-ingestion

Retr0Metal98 committed:
Delete lta_camera.py

Superseded by ingest.py

Retr0Metal98 committed:
Synchronizing weather data ingestion

Weather data now includes a field `call_timestamp` - when the API was called.
- This can be used to more easily join the tables together.
- enables all APIs to be called using the same timestamp parameter to synchronize.

The merging of csv files will not drop duplicate rows by default .
- this is so that weather info provided at an earlier time won't be overwritten by updated info.
- camera images retrieved at a given time can be merged to the latest weather info available *at that time*
- downstream table joins and transforms should account for this

Retr0Metal98 committed:
Update weather_data.py

corrected date_time parameter for weather API calls

don-tay committed:
Add image ingest into scheduled run
